src/v/cluster/scheduling/partition_allocator.cc

```diff
@@ -72,13 +85,177 @@ partition_allocator::allocate_partition(partition_constraints p_constraints) {
+    uint64_t effective_cluster_memory = all_brokers.size() * min_memory_bytes;
+    uint64_t effective_cluster_disk = all_brokers.size() * min_disk_bytes;
+
+    vlog(
+      clusterlog.debug,
+      "Effective cluster resources: CPU={} memory={} disk={}",
+      effective_cpu_count,
+      human::bytes(effective_cluster_memory),
+      human::bytes(effective_cluster_disk));
+
+    // Refuse to create a partition count that would violate the per-core
+    // limit.
+    const uint64_t core_limit = effective_cpu_count
+                                * allocation_node::max_allocations_per_core;
+    if (proposed_total_partitions > core_limit) {
+        vlog(
+          clusterlog.warn,
+          "Refusing to create {} partitions, exceeds core limit {}",
+          create_count,
+          effective_cpu_count * allocation_node::max_allocations_per_core);
+        return errc::topic_invalid_partitions;
+    }
+
+    // Refuse to create partitions that would violate the configured
+    // memory per partition.
+    auto memory_per_partition_replica = _memory_per_partition();
+    if (memory_per_partition_replica.has_value()) {
+        const uint64_t memory_limit = effective_cluster_memory
+                                      / memory_per_partition_replica.value();
+
+        if (memory_limit > 0 && proposed_total_partitions > memory_limit) {
+            vlog(
+              clusterlog.warn,
+              "Refusing to create {} partitions, exceeds memory limit {}",
+              create_count,
+              memory_limit);
+            return errc::topic_invalid_partitions;
+        }
+    }
+
+    // Refuse to create partitions that would exhaust our nfiles ulimit
+    auto fds_per_partition_replica = _fds_per_partition();
+    if (fds_per_partition_replica.has_value()) {
+        struct rlimit nofile = {0, 0};
+        if (getrlimit(RLIMIT_NOFILE, &nofile) == 0) {
+            if (nofile.rlim_cur != RLIM_INFINITY) {
+                const uint64_t fds_limit = (all_brokers.size()
+                                            * nofile.rlim_cur)
+                                           / fds_per_partition_replica.value();
+                if (proposed_total_partitions > fds_limit) {
+                    vlog(
+                      clusterlog.warn,
+                      "Refusing to create {} partitions, exceeds FD limit {}",
+                      create_count,
+                      fds_limit);
+                    return errc::topic_invalid_partitions;
+                }
+            }
+        } else {
+            // Unclear if this can ever happen, but let's be graceful.
+            vlog(clusterlog.warn, "Error {} querying file handle limit", errno);
+        }
+    }
+
+    // Refuse to create partitions if there isn't enough space to at least
+    // falloc the first part of a segment for each partition
+    uint64_t disk_limit = effective_cluster_disk / _fallocation_step();
```

Comment:
Is division by zero possible here?

Reply:
bounds on fallocation step size are enforced by configuration validation (see `storage::segment_appender::validate_fallocation_step`)